core/tests/test_utils.py

```diff
@@ -1838,6 +1843,92 @@ def swap(self, obj, attr, newvalue):
         finally:
             setattr(obj, attr, original)
 
+    @contextlib.contextmanager
+    def swap_with_checks(
+            self, obj, attr, newvalue, expected_args=None, expected_kwargs=None,
+            called=True):
+        """Swap an object's function value within the context of a
+        'with' statement. The object can be anything that supports
+        getattr and setattr, such as class instances, modules, ...
+        Yields:
+            context: The context with function replaced.
+
+        Examples:
+            If you want to check subprocess.Popen is invoked twice
+            like `subprocess.Popen(['python'], shell=True)` and
+            `subprocess.Popen(['python2], shell=False), you can first
+            define the mock function, then the swap, and just run the
+            target function in context, as follows:
+                def mock_popen(command, shell):
+                    return
+
+                popen_swap = self.swap_with_checks(
+                    subprocess, 'Popen', mock_popen, expected_args=[
+                        (['python'],), (['python2'],)], expected_kwargs=[
+                            {'shell': True,}, {'shell': False}])
+                with popen_swap:
+                    function_that_invokes_popen()
+
+        Args:
+            obj: *. the Python object whose attribute you want to swap.
+            attr: str. The name of the function to be swapped.
+            newvalue: Function. The new function you want to use.
+            expected_args: None|list(tuple). The expected args that you
+                want this function to be invoked with. When its value is None,
+                args will not be checked.If the value type is list, the function
+                will check whether the called args is the first element in the
+                list. If matched, this tuple will be removed from the list.
+            expected_kwargs: None|list(dict). The expected key word args
+                you want this function to be invoked with. Similar to
+                expected_args.
+            called: bool. Whether the function is invoked. This will always be
+                checked.
+        """
+        original = getattr(obj, attr)
+        msg = 'Failed in check function %s.%s' % (obj.__name__, attr)
```

Comment:
This message is only designed to tell the user which function failed the validation. 

The detailed information, e.g.: the original value and expected value, is added by the `assert` functions - that is also why I added the `self.longMessage = True`. By default there is no origin and expected value print out by the assert function so I need to specify this value to be `True` to tell it to do so. 

I switched the value back to `False` later on because I do not want to change the other `assert`'s behaviours. 

Here is an example of the actual output:
```
AssertionError: {'args': []} != {u'args': [1, 2, 3]}
- {'args': []}
+ {u'args': [1, 2, 3]}
?  +         +++++++
 : Failed in check function scripts.setup_gae.main
```

Reply:
Ah, ok. That sounds fine to me, maybe add a comment in the code to explain that "this message will be printed after the AssertionError debug logs to explain which function failed the validation".

But, looking at your message, I'm still a bit confused about where to look (as a developer) -- where is the "check function scripts.setup_gae.main"? Will there be a stacktrace that I can use to debug? Also I am not sure what the relevance of scripts.setup_gae.main is, here -- could you explain?

(Maybe we can rephrase it to something like: "Expected checks failed when swapping out scripts.setup_gae.main in tests." -- if that is what it actually means.)

Reply:
Thanks, that is what I mean. 
There is a stack trace in the error message.